### DISCUSSION TITLE
Is there an option for screen capture and recording?

### DISCUSSION THREAD

**Discussion by wnsntgs** [microsoft/WindowsAppSDK#2645] [Q&A]
I am interested in the ability to capture a screen (screenshot), as well as to record a screen (as well as an arbitrary part of the screen) with and without sound (without using the screen recording picker from UWP).

**Comment by dhoehna** [comment]
Hi wnstngs,

Are you wondering if Windows App SDK has an option to record the screen, or are you wondering how to, in general, record the screen, or something else?

**Reply by wnsntgs** [reply]
I know how to do it using 'classic' Windows API and GDI. Using WinUI 3 I don't know how to do this and if it is even possible to capture and record the screen. I know that UWP (WinUI 2) has a way to record screen or window, but only using their API with a ready-made picker. I don't like this way and I wonder if WinUI 3 offers more freedom in this matter.

**Reply by dhoehna** [reply]
I see.  Sorry, I didn't understand that you were looking for a way to do this in WinUi3.  @MikeHillberg does WinUI3 have an API to capture the screen?

**Reply by MikeHillberg** [reply]
I think it's possible with [GraphicsCaptureItem.TryCreateFromWindowId](https://docs.microsoft.com/uwp/api/Windows.Graphics.Capture.GraphicsCaptureItem.TryCreateFromWindowId), but it's complicated today ... You can get a WindowID from an hwnd with [GetWindowIdFromWindow](https://docs.microsoft.com/en-us/windows/apps/api-reference/cs-interop-apis/microsoft.ui/microsoft.ui.win32interop.getwindowidfromwindow), and get an hwnd from a Xaml Window with [IWindowNative](https://docs.microsoft.com/en-us/windows/windows-app-sdk/api/win32/microsoft.ui.xaml.window/nn-microsoft-ui-xaml-window-iwindownative).

**Reply by unboundinnov** [reply]
@MikeHillberg do you have any links to samples for using these methods?

**Comment by castorix** [comment]
There are many ways.
GDI, GDI+, DirectX ([D3DXSaveSurfaceToFile](https://docs.microsoft.com/en-us/windows/win32/direct3d9/d3dxsavesurfacetofile) or D3D11), ... for a simple Screen Capture (can be used with a WriteableBitmap), 
[Sink Writer](https://docs.microsoft.com/en-us/windows/win32/medfound/sink-writer) for video capture,
Desktop Duplication API, ...

**Reply by rcohn** [reply]
Hello @castorix -

I have been searching for an API to capture screen images in a WinUI 3 app (without the user selecting from a picker).  The responses in this thread seem to indicate that such exists.  My goal is to capture screen images under programmatic control and display them on a Canvas element.  Most of what I've found so far are APIs that work only on Windows forms.

Could you provide some examples or other pointers showing screen capture and display using one of the APIs you name above (GDI, GDI+, ...)?

Very much appreciated.  Thank you.

**Reply by castorix** [reply]
> Hello @castorix -
> 
> I have been searching for an API to capture screen images in a WinUI 3 app (without the user selecting from a picker). The responses in this thread seem to indicate that such exists. My goal is to capture screen images under programmatic control and display them on a Canvas element. Most of what I've found so far are APIs that work only on Windows forms.
> 
> Could you provide some examples or other pointers showing screen capture and display using one of the APIs you name above (GDI, GDI+, ...)?
> 
> Very much appreciated. Thank you.

As I posted below (message on Oct 16, 2022), the _Desktop Duplication_ method seems to work fine
I have a (quick & dirty...) test code that I can clean to upload it on Github.

**Reply by rcohn** [reply]
Hi @castorix -

That would be great if you can post your "quick & dirty" Desktop Duplication test code.  I would be extremely happy to try it out.

Thanks very much.

**Reply by castorix** [reply]
> Hi @castorix -
> 
> That would be great if you can post your "quick & dirty" Desktop Duplication test code. I would be extremely happy to try it out.
> 
> Thanks very much.

I just  uploaded a test sample : [WinUI3_DesktopDuplication](https://github.com/castorix/WinUI3_DesktopDuplication/tree/master)
 I used a Win32 [Hot Key Control]( https://learn.microsoft.com/en-us/windows/win32/controls/hot-key-controls) for the test, but Win32 controls should not be mixed with XAML controls. I used it to simplify the code to manage Hot keys.
 You must set the focus in this control and choose a Hot key, like [Ctrl] + F12 for example (F12 alone is prohibited)
 Then you can minimize the app and when you will hit the Hot key, it will do a screen copy (copied in a _WriteableBitmap_ then in an _Image_ control) and set the window to foreground
Only tested on Windows 10 (22H2)

**Reply by rcohn** [reply]
Hi @castorix -

I built and ran the demo code.  As I learned, it captures the image on the first enumerated display.  I experimented by having it capture images for my other two displays.  Most of the time, the demo presented the image quickly without problems.  There were times when I started the app up that it would crash the first time I hit the hotkey (CTL).  Then I would bring it up again and it worked great.  I'm thinking that some GXDI thing had not been fully initialized and that adding safeguard code would probably help there.  I built and ran the demo on my Win11 machine.

I'm really not surprised by the complexity of the solution that is needed (ie, interop to the Win32 IDXGI interface) since there is no WinUI3-compatible .NET API to perform this capture.  I don't claim by any stretch that I understand all of the DX complexities (yet), but I can follow the string of operations in CaptureFrame to the point where the image source gets set.  That is very useful.

Once again, many thanks for the great work on this and keeping watch on so many other questions in this forum.

Best regards.

**Reply by castorix** [reply]
> There were times when I started the app up that it would crash the first time I hit the hotkey (CTL).

Could you catch the error message or the call stack ?
I probably made a mistake somewhere for a particular case...

**Reply by rcohn** [reply]
I will try to send you back some diagnostic info - possibly not before Monday. 

On a related note, and please forgive my ignorance, but I have a question about the interface GUIDs.  It looks like these are possibly documented in https://learn.microsoft.com/en-us/windows/win32/direct3ddxgi/d3d10-graphics-reference-dxgi or some related site.  In production code, is it normally the case that the GUIDs are explicitly written out for these interfaces as in your demo code, or can defines be used?  I realize that MS cannot (or should not) arbitrarily change these GUIDs as it would break a lot of existing code.  Just wondering.  Thanks.

**Reply by castorix** [reply]
For all the P/Invoke declarations, I always copy/translate them into C# from the [Windows SDK](https://developer.microsoft.com/en-us/windows/downloads/windows-sdk/) C++ headers
The GUID constants never change when they are documented in MSDN and are in SDK
For example, for the first interface in DXGI.cs, _IDXGIObject_, I translated it from _dxgi.h_ :

    MIDL_INTERFACE("aec22fb8-76f3-4639-9be0-28eb43a67a2e")
    IDXGIObject : public IUnknown
    {
    public:
        virtual HRESULT STDMETHODCALLTYPE SetPrivateData( 
            /* [annotation][in] */ 
            _In_  REFGUID Name,
            /* [in] */ UINT DataSize,
            /* [annotation][in] */ 
            _In_reads_bytes_(DataSize)  const void *pData) = 0;
        
        virtual HRESULT STDMETHODCALLTYPE SetPrivateDataInterface( 
            /* [annotation][in] */ 
            _In_  REFGUID Name,
            /* [annotation][in] */ 
            _In_opt_  const IUnknown *pUnknown) = 0;
        
        virtual HRESULT STDMETHODCALLTYPE GetPrivateData( 
            /* [annotation][in] */ 
            _In_  REFGUID Name,
            /* [annotation][out][in] */ 
            _Inout_  UINT *pDataSize,
            /* [annotation][out] */ 
            _Out_writes_bytes_(*pDataSize)  void *pData) = 0;
        
        virtual HRESULT STDMETHODCALLTYPE GetParent( 
            /* [annotation][in] */ 
            _In_  REFIID riid,
            /* [annotation][retval][out] */ 
            _COM_Outptr_  void **ppParent) = 0;
        
    };

**Reply by rcohn** [reply]
Here is the exception output when it crashes (I tried to paste into this reply, but the formatting looked all wrong):

[WinUI3_DesktopDuplication_exception.txt](https://github.com/microsoft/WindowsAppSDK/files/10636754/WinUI3_DesktopDuplication_exception.txt)

The exception is on line 241: 
`pD3D11DeviceContext.CopyResource(m_textureGDI, pD3D11Texture2D);`

It seems that crashes mostly occur when I try to do the capture within about 10 seconds of starting the app.  If I wait 10-15 seconds, it usually works (but not always).

BTW - I also tested on Win10 and the behavior was mostly the same.

**Reply by castorix** [reply]
I tried dozens of times to do the capture within the first 1/2 seconds, but impossible to reproduce...
 (I have an old PC without real graphic card (_Intel(R) HD Graphics)_)
 
 The code could be improved by testing if _m_textureGDI_ has been created in
 
 _hr = pD3D11Device.CreateTexture2D(textureDesc, IntPtr.Zero, out m_textureGDI);_
 
 by adding a Dialog Box if (hr != HRESULT.S_OK) inside Initialize();
 
 like, at the end of the function
 
            if (hr != HRESULT.S_OK)
            {
                string sError = "Could not initialize D3D11" + "\r\n" + "HRESULT = 0x" + string.Format("{0:X}", hr);
                Windows.UI.Popups.MessageDialog md = new Windows.UI.Popups.MessageDialog(sError, "Error");
                WinRT.Interop.InitializeWithWindow.Initialize(md, hWnd);
                _ = await md.ShowAsync();
            }
            return hr;
            
 + renaming
            
            private async Task<HRESULT> InitializeAsync()
            
 + modifying the call at beginning :
            
           Task<HRESULT> hr = InitializeAsync();
 
 and inside CaptureFrame, adding a try/catch or testing if _m_textureGDI_ != null :
 
```
   if (m_textureGDI != null)
   {
       pD3D11DeviceContext.CopyResource(m_textureGDI, pD3D11Texture2D);
       // all the code...
   }
```

**Reply by rcohn** [reply]
I added the above changes and the result is the same.  Also put a breakpoint in the check for hr != HRESULT.S_OK and it never breaks there.  Funny thing is I can rerun the app several times in a row and it works fine.  And then, I try again and it crashes on the same CopyResource call.

Thanks.

**Reply by castorix** [reply]
> I added the above changes and the result is the same. Also put a breakpoint in the check for hr != HRESULT.S_OK and it never breaks there. Funny thing is I can rerun the app several times in a row and it works fine. And then, I try again and it crashes on the same CopyResource call.

You mean it does not reach this BP inside _private async Task<HRESULT> InitializeAsync()_ ?

![InitializeAsync](https://user-images.githubusercontent.com/22345506/217078528-a53f8df5-d51f-45a6-aa83-a7121c831ae1.jpg)

From what I understand, _m_textureGDI_ is never null when it crashes ?
Did you try to put at least the line ;
 _pD3D11DeviceContext.CopyResource(m_textureGDI, pD3D11Texture2D);_
inside a try/catch to see if it can be intercepted ?
+
adding PreserveSig in the declaration of CopyResource :
```
       [PreserveSig]
        void CopyResource(ID3D11Resource pDstResource, ID3D11Resource pSrcResource);
```
    
Something like this (or without MessageDialog, if the window is not in foreground... just a System.Diagnostics.Debug.WriteLine to display the _sError_  string) :

                    try
                    {
                        pD3D11DeviceContext.CopyResource(m_textureGDI, pD3D11Texture2D);                                             
                    }
                    catch (Exception ex)
                    {
                        string sError = ex.Message + "\r\n" + "HRESULT = 0x" + string.Format("{0:X}", ex.HResult);
                        Windows.UI.Popups.MessageDialog md = new Windows.UI.Popups.MessageDialog(sError, "Error");
                        WinRT.Interop.InitializeWithWindow.Initialize(md, hWnd);
                        _ = await md.ShowAsync();
                    }
                    finally
                    {
                        SafeRelease(ref pD3D11Texture2D);
                        SafeRelease(ref pD3D11DeviceContext);
                        hr = m_pDXGIOutputDuplication.ReleaseFrame();
                    }

**Reply by rcohn** [reply]
I have not seen m_textureGDI be null.  I added all of the above changes up to PreserveSig and it was still exceptioning in the same place.  Then I added the PreserveSig and it started to work reliably every time.  However, that seems to be a coincidence because it still worked fine after I stubbed that out.  Anyway, will try again later.  Will let you know what happens.

**Reply by castorix** [reply]
The resource seems to have the _D3D11_RESOURCE_MISC_SHARED_KEYEDMUTEX_ flag
and apparently from MSDN, I would need to call _IDXGIKeyedMutex::AcquireSync_ and _ReleaseSync_ APIs.
I will try to add them, but not easy to debug as I cannot reproduce the crash.

**Reply by rcohn** [reply]
Since the last set of changes added, I have not been able to reproduce the crash myself (at least on Win11, didn't try Win10 yet).  However, if there is a D3Dll mutex being used, it seems that having the app acquire and release it is a safe (and advisable) thing to do.

If you put more changes in places, I will give them a try.

Thanks.

**Reply by castorix** [reply]
ATM, my first tries with _IDXGIKeyedMutex_ ... are a failure.
I did some tests from MSDN doc at https://learn.microsoft.com/en-us/windows/win32/direct3darticles/surface-sharing-between-windows-graphics-apis
but I get locks or black texture or crashes...

**Reply by rcohn** [reply]
I was shown a way to use System.Drawing.Graphics.CopyFromScreen(), which works quite well.  I use that method to produce a System.Drawing.Bitmap, which I can then use with Bitmap.Save to produce a png format image file.  I save the PNG to the user's temp space, from where I can read it back in and use to Source an image.  There is probably a way to do that using a stream, which I will have to investigate when I get some time.

Anyway, this works fine for my needs. Thanks again for your help.

**Reply by castorix** [reply]
Yes, CopyFromScreen is the old GDI method with BitBlt(), that can work for most cases

**Comment by khoabui1412** [comment]
I'm trying to use GraphicsCaptureItem to capture screen but: `GraphicsCaptureItem.TryCreateFromWindowId` only supports on win11.
I use this instead:
```
        static readonly Guid GraphicsCaptureItemGuid = new Guid("79C3F95B-31F7-4EC2-A464-632EF5D30760");
        [ComImport]
        [Guid("3628E81B-3CAC-4C60-B7F4-23CE0E0C3356")]
        [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
        [ComVisible(true)]
        interface IGraphicsCaptureItemInterop
        {
            IntPtr CreateForWindow(
                [In] IntPtr window,
                [In] ref Guid iid);

            IntPtr CreateForMonitor(
                [In] IntPtr monitor,
                [In] ref Guid iid);
        }
        public static GraphicsCaptureItem CreateItemForWindow(IntPtr hwnd)
       {
            var interop = GraphicsCaptureItem.As<IGraphicsCaptureItemInterop>();
            var itemPointer = interop.CreateForWindow(hwnd, GraphicsCaptureItemGuid);
            var item = Marshal.GetObjectForIUnknown(itemPointer) as GraphicsCaptureItem;
            Marshal.Release(itemPointer);
            return item:
       }
```
But the item is null, am I missing something? is it possibile on win10? Thanks

**Reply by castorix** [reply]
> is it possibile on win10?

If you only want to capture screen, you can just use GDI (a simple BitBlt for example) or to capture a window, PrintWindow to a memory DC (with PW_RENDERFULLCONTENT flag)

**Reply by MathewRimmington** [reply]
I had the same problem, not sure if you ever resolved this, I implemented this myself swapping line
`var item = Marshal.GetObjectForIUnknown(itemPointer) as GraphicsCaptureItem;`

with
`var item = WinRT.MarshalInterface<GraphicsCaptureItem>.FromAbi(itemPointer);`

**Reply by castorix** [reply]
> [MathewRimmington](https://github.com/MathewRimmington)
> [2 days ago](https://github.com/microsoft/WindowsAppSDK/discussions/2645?sort=new#discussioncomment-9481064)
> I had the same problem, not sure if you ever resolved this, I implemented this myself swapping line
> var item = Marshal.GetObjectForIUnknown(itemPointer) as GraphicsCaptureItem;
> with
> var item = WinRT.MarshalInterface<GraphicsCaptureItem>.FromAbi(itemPointer);

It works on my configuration (Windows 10 22H2, Windows App SDK 1.4.2)

**Comment by castorix** [comment]
I quickly tested all the methods and Desktop Duplication ([IDXGIOutputDuplication::AcquireNextFrame](https://learn.microsoft.com/en-us/windows/win32/api/dxgi1_2/nf-dxgi1_2-idxgioutputduplication-acquirenextframe)) converted to a _WriteableBitmap_ works fine (better than _BitBlt_ which cannot capture DirectX)
I could upload a test sample if you cannot make _GraphicsCaptureItem_ work (I also get a null value from _Marshal.GetObjectForIUnknown_ on Windows 10 21H1)

**Reply by sigmarsson** [reply]
Could you share the snippet, which is working on 10 and/or 11 ? Merci!
